pkg/scheduler/objects/application.go

```diff
@@ -1103,11 +1177,19 @@ func (sa *Application) removeAllocationInternal(uuid string) *Allocation {
 		sa.allocatedResource = resources.Sub(sa.allocatedResource, alloc.AllocatedResource)
 	}
 	// When the resource trackers are zero we should not expect anything to come in later.
-	if resources.IsZero(sa.pending) && resources.IsZero(sa.allocatedResource) && resources.IsZero(sa.allocatedPlaceholder) {
-		if err := sa.HandleApplicationEvent(WaitApplication); err != nil {
-			log.Logger().Warn("Application state not changed to Waiting while removing some allocation(s)",
-				zap.String("currentState", sa.CurrentState()),
-				zap.Error(err))
+	if resources.IsZero(sa.pending) && resources.IsZero(sa.allocatedResource) {
+		if sa.isWaitingStateTimedOut() && resources.IsZero(sa.allocatedPlaceholder) {
+			if err := sa.HandleApplicationEvent(CompleteApplication); err != nil {
+				log.Logger().Warn("Application state not changed to Completed while removing some allocation(s)",
+					zap.String("currentState", sa.CurrentState()),
+					zap.Error(err))
+			}
+		} else {
+			if err := sa.HandleApplicationEvent(WaitApplication); err != nil {
+				log.Logger().Warn("Application state not changed to Waiting while removing some allocation(s)",
+					zap.String("currentState", sa.CurrentState()),
+					zap.Error(err))
+			}
```

Comment:
The reason we do not do that is because an application could crash or fail before the placeholder time out expires. If we do not handle that we could leave unusable placeholders on the nodes. That would mean we are holding on to resources which we will not use and could use for other things.
For instance: a spark driver crashes at startup. In that case we have no real pods running and there will never be a pod running as nothing will request the executors. We need to clean up what is left.

Reply:
We had a sync with @yangwwei and we decided to simplify this one as he suggested, and after we have some feedback we can enhance it in the future.